### ISSUE TITLE
Linking error MacOS gcc-10

### ISSUE THREAD

**Issue by mircodemarchi** [mlpack/mlpack#2934]
MacOS version: macOS Catalina 10.15.7

GCC version: Homebrew GCC 10.2.0_4

I have installed mlpack with `brew install mlpack` and I'm trying to build with g++-10 (the command used: `g++-10 -std=c++11 test.cpp -lmlpack -larmadillo`) the following source file (test.cpp): 

```
#include <mlpack/core.hpp>

using namespace mlpack;

int main() {
    arma::mat data;
    data::Load("data/execution-time.csv", data, true);
}
```

The error returned is the following: 

```
Undefined symbols for architecture x86_64:
  "__ZN6mlpack4data4LoadIdEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERN4arma3MatIT_EEbbNSA_9file_typeE", referenced from:
      _main in ccQnUBlY.o
  "___emutls_v._ZN4arma19mt19937_64_instanceE", referenced from:
      __ZTWN4arma19mt19937_64_instanceE in ccQnUBlY.o
ld: symbol(s) not found for architecture x86_64
collect2: error: ld returned 1 exit status
```

If I try with clang compiler, it works, but I need to build a library really portable, that works in different operating systems and environments, therefore I need the compatibility with gcc-10. I tried also with Ubuntu 20.04 and mlpack works with both gcc and clang. I can't explain why in ubuntu works and in macOS not :(

In addition, the environments that I need are reproduced with GitHub Actions Continuous Integration in this [repository](https://github.com/mircodemarchi/AriadneDL). You can see that the project works correctly in every environment except one, that is macOS with gcc-10. If you select the build that fails, you can also get additional logs. 

Thank you in advance.

**Comment by zoq** [comment]
I haven't looked into the brew formula but I guess they use `stdlib=libstdc++` instead of `-stdlib=libc++` so the output binaries are not compatible with gcc. Instead of installing mlpack through brew can you build mlpack locally and use `gcc` instead e.g. by setting `CXX` and `CC`?

**Comment by mircodemarchi** [comment]
Thank you very much for the answer,
yes, I can install mlpack locally, I will try and I let you know. 

However, since my project will be an extension of another library, I would like my final project to be available without the user installing mlpack locally, but instead using homebrew to be able to install it easily. 
When will this fix also be available on homebrew?

**Comment by zoq** [comment]
Hm, looking at the CMake file:

https://github.com/mlpack/mlpack/blob/596859cb5bd7217a2f159097e1edb9d5a7f2ae83/CMakeLists.txt#L176-L181

should handle this, so I guess let's see if that solves the issue you see and if it does, I'll take a closer look into the issue.

**Comment by mircodemarchi** [comment]
The CMakeList file is correct, because it sets the `-stdlib=libc++` flag if the clang compiler is used.
https://github.com/mlpack/mlpack/blob/596859cb5bd7217a2f159097e1edb9d5a7f2ae83/CMakeLists.txt#L162
In fact, I found that `-stdlib=libc++` flag is used only in clang ([here](https://stackoverflow.com/questions/24419832/unrecognized-command-line-option-stdlib-libc-with-macports-gcc48)).

However, the problem (that I have discussed [above](https://github.com/mlpack/mlpack/issues/2934#issue-870069809)) is related with gcc-10, that is the GNU version of gcc installed through Homebrew in macOS.

Anyhow, I tried to install mlpack from source and the compilation process finishes with a linking error similar to the one showed above. The steps that I follow to install mlpack are the following: 
- wget https://www.mlpack.org/files/mlpack-3.4.2.tar.gz
- tar -xvzpf mlpack-3.4.2.tar.gz
- mkdir mlpack-3.4.2/build && cd mlpack-3.4.2/build
- cmake .. -DCMAKE_C_COMPILER=gcc-10 -DCMAKE_CXX_COMPILER=g++-10
- make -j4

The following is cmake command output:
```
-- The C compiler identification is GNU 10.2.0
-- The CXX compiler identification is GNU 10.2.0
-- Checking whether C compiler has -isysroot
-- Checking whether C compiler has -isysroot - yes
-- Checking whether C compiler supports OSX deployment target flag
-- Checking whether C compiler supports OSX deployment target flag - yes
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/local/bin/gcc-10 - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Checking whether CXX compiler has -isysroot
-- Checking whether CXX compiler has -isysroot - yes
-- Checking whether CXX compiler supports OSX deployment target flag
-- Checking whether CXX compiler supports OSX deployment target flag - yes
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/local/bin/g++-10 - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- cotire 1.7.10 loaded.
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Found Armadillo: /usr/local/lib/libarmadillo.dylib (found suitable version "10.4.1", minimum required is "8.400.0")
-- Could NOT find StbImage (missing: STB_IMAGE_INCLUDE_DIR)
-- [download 100% complete]
* Closing connection 0
-- [download 100% complete]
* Closing connection 0
CMake Warning at CMakeLists.txt:340 (message):
  Could not download stb! Error code : "HTTP response code said error"! Error
  log:


CMake Warning at CMakeLists.txt:342 (message):
  stb/stb_image.h is not installed.  Image utilities will not be available!


-- Found Ensmallen: /usr/local/include (found suitable version "2.16.2", minimum required is "2.10.0")
-- Found Boost: /usr/local/include (found suitable version "1.75.0", minimum required is "1.58") found components: unit_test_framework serialization
-- Found OpenMP_C: -fopenmp (found version "4.5")
-- Found OpenMP_CXX: -fopenmp (found version "4.5")
-- Found OpenMP: TRUE (found version "4.5")
-- Found Git: /usr/bin/git (found version "2.24.3 (Apple Git-128)")
-- Regenerating arma_config.hpp.
CMake Warning at CMakeLists.txt:585 (message):
  txt2man not found; man pages will not be generated.


-- Julia not found (missing: JULIA_EXECUTABLE) (Required is at least version "0.7.0")
-- Julia not found; not building Julia bindings.
-- Not building Markdown bindings.
-- Not building Python bindings.
-- Go not found (missing: GO_EXECUTABLE) (Required is at least version "1.11.0")
-- Gonum not found (missing: GONUM_VERSION_STRING)
-- Not building Go bindings; the following modules are not available:
    - Go
    - Gonum
-- R not found (missing: R_EXECUTABLE) (Required is at least version "4.0")
-- Could NOT find R_roxygen2 (missing: R_ROXYGEN2)
-- Could NOT find R_Rcpp (missing: R_RCPP)
-- Could NOT find R_RcppArmadillo (missing: R_RCPPARMADILLO)
-- Could NOT find R_RcppEnsmallen (missing: R_RCPPENSMALLEN)
-- Could NOT find R_BH (missing: R_BH)
-- Could NOT find R_testthat (missing: R_TESTTHAT)
-- Not building R bindings, Following modules are not available:
    - R
    - roxygen2
    - Rcpp
    - RcppArmadillo
    - RcppEnsmallen
    - BH
    - testthat
-- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY
-- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY - Success
-- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY
-- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY - Success
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR - Success
-- CXX target mlpack cotired.
-- CXX target mlpack_test cotired.
-- CXX target mlpack_catch_test cotired.
CMake Warning (dev) at /usr/local/Cellar/cmake/3.20.2/share/cmake/Modules/GNUInstallDirs.cmake:357 (message):
  GNUInstallDirs_get_absolute_install_dir called without third argument.
  Using ${dir} from the caller's scope for compatibility with CMake 3.19 and
  below.
Call Stack (most recent call first):
  src/mlpack/CMakeLists.txt:166 (GNUInstallDirs_get_absolute_install_dir)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at /usr/local/Cellar/cmake/3.20.2/share/cmake/Modules/GNUInstallDirs.cmake:357 (message):
  GNUInstallDirs_get_absolute_install_dir called without third argument.
  Using ${dir} from the caller's scope for compatibility with CMake 3.19 and
  below.
Call Stack (most recent call first):
  src/mlpack/CMakeLists.txt:167 (GNUInstallDirs_get_absolute_install_dir)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found Doxygen: /usr/local/bin/doxygen (found version "1.9.1") found components: doxygen missing components: dot
-- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE)
-- Configuring done
CMake Warning (dev) at CMake/cotire.cmake:3048 (add_library):
  Policy CMP0115 is not set: Source file extensions must be explicit.  Run
  "cmake --help-policy CMP0115" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  File:

    /Users/mircodemarchi/Repositories/mlpack-3.4.2/src/mlpack/methods/ann/gan/metrics/inception_score.hpp
Call Stack (most recent call first):
  CMake/cotire.cmake:3261 (cotire_setup_unity_build_target)
  CMake/cotire.cmake:3431 (cotire_target)
  src/mlpack/CMakeLists.txt:70 (cotire)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at CMake/cotire.cmake:3048 (add_library):
  Policy CMP0115 is not set: Source file extensions must be explicit.  Run
  "cmake --help-policy CMP0115" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  File:

    /Users/mircodemarchi/Repositories/mlpack-3.4.2/src/mlpack/methods/ann/gan/metrics/inception_score_impl.hpp
Call Stack (most recent call first):
  CMake/cotire.cmake:3261 (cotire_setup_unity_build_target)
  CMake/cotire.cmake:3431 (cotire_target)
  src/mlpack/CMakeLists.txt:70 (cotire)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Generating done
-- Build files have been written to: /Users/mircodemarchi/Repositories/mlpack-3.4.2/build
```

The following is build process `make` output:
```
[  0%] Updating arma_config.hpp (if necessary)
[  0%] Built target mlpack_arma_config
Moving header files to include/mlpack/
[  0%] Built target mlpack_headers
[  0%] Generating CXX prefix source src/mlpack/cotire/mlpack_CXX_prefix.cxx
[  0%] Generating CXX prefix header src/mlpack/cotire/mlpack_CXX_prefix.hxx
[  1%] Building CXX precompiled header src/mlpack/cotire/mlpack_CXX_prefix.hxx.gch
Scanning dependencies of target mlpack
[  2%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/data/load_csv.cpp.o
[  2%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/bindings/cli/print_help.cpp.o
[  2%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/data/detect_file_type.cpp.o
[  2%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/bindings/tests/clean_memory.cpp.o
[  3%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/data/load_image.cpp.o
[  3%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/data/load.cpp.o
[  3%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/data/save_image.cpp.o
[  3%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/dists/discrete_distribution.cpp.o
[  4%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/dists/gaussian_distribution.cpp.o
[  4%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/dists/laplace_distribution.cpp.o
[  4%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/dists/regression_distribution.cpp.o
[  5%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/dists/gamma_distribution.cpp.o
[  5%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/dists/diagonal_gaussian_distribution.cpp.o
[  5%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/kernels/epanechnikov_kernel.cpp.o
[  5%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/kernels/pspectrum_string_kernel.cpp.o
[  6%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/math/columns_to_blocks.cpp.o
[  6%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/math/lin_alg.cpp.o
[  6%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/math/random.cpp.o
[  7%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/math/random_basis.cpp.o
[  7%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/tree/cosine_tree/cosine_tree.cpp.o
[  7%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/util/backtrace.cpp.o
[  7%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/util/io.cpp.o
[  8%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/util/log.cpp.o
[  8%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/util/prefixedoutstream.cpp.o
[  8%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/util/program_doc.cpp.o
[  9%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/util/singletons.cpp.o
[  9%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/util/timers.cpp.o
[  9%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/util/version.cpp.o
[  9%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/adaboost/adaboost_model.cpp.o
[ 10%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/bayesian_linear_regression/bayesian_linear_regression.cpp.o
[ 10%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/block_krylov_svd/randomized_block_krylov_svd.cpp.o
[ 10%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/fastmks/fastmks_model.cpp.o
[ 11%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/gmm/gmm.cpp.o
[ 11%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/gmm/diagonal_gmm.cpp.o
[ 11%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/hoeffding_trees/hoeffding_tree_model.cpp.o
[ 11%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/lars/lars.cpp.o
[ 12%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/linear_regression/linear_regression.cpp.o
[ 12%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/local_coordinate_coding/lcc.cpp.o
[ 12%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/matrix_completion/matrix_completion.cpp.o
[ 13%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/neighbor_search/unmap.cpp.o
[ 13%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/quic_svd/quic_svd.cpp.o
[ 13%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/radical/radical.cpp.o
[ 13%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/randomized_svd/randomized_svd.cpp.o
[ 15%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/rann/ra_util.cpp.o
[ 15%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/softmax_regression/softmax_regression.cpp.o
[ 15%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/softmax_regression/softmax_regression_function.cpp.o
[ 16%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/sparse_autoencoder/sparse_autoencoder.cpp.o
[ 16%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/sparse_autoencoder/sparse_autoencoder_function.cpp.o
[ 16%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/sparse_autoencoder/maximal_inputs.cpp.o
[ 16%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/sparse_coding/sparse_coding.cpp.o
[ 17%] Linking CXX shared library ../../lib/libmlpack.dylib
Undefined symbols for architecture x86_64:
  "___emutls_v._ZN4arma19mt19937_64_instanceE", referenced from:
      __GLOBAL__sub_I_print_help.cpp in print_help.cpp.o
      __GLOBAL__sub_I_clean_memory.cpp in clean_memory.cpp.o
      __GLOBAL__sub_I_detect_file_type.cpp in detect_file_type.cpp.o
      __GLOBAL__sub_I_load_csv.cpp in load_csv.cpp.o
      __GLOBAL__sub_I_load_image.cpp in load_image.cpp.o
      __GLOBAL__sub_I_load.cpp in load.cpp.o
      __GLOBAL__sub_I_save_image.cpp in save_image.cpp.o
      ...
ld: symbol(s) not found for architecture x86_64
collect2: error: ld returned 1 exit status
make[2]: *** [lib/libmlpack.3.4.dylib] Error 1
make[1]: *** [src/mlpack/CMakeFiles/mlpack.dir/all] Error 2
make: *** [all] Error 2
```

I tried the same steps using clang (i.e., omitting compiler args in cmake command) and the process ends without errors.

**Comment by RishabhGarg108** [comment]
Hey @mircodemarchi , I also faced a somewhat similar issue with `g++-10` and `clang`. But, you can use `g++` and that can probably be a workaround for this issue.

What you can do is, build mlpack from the source following the instructions in README.md. In the cmake command, do not force it to use g++-10. Let it use the default one (which I guess is g++). Once mlpack is installed, then for compiling your mlpack programs, you can use the following command:

`g++ -std=c++11 -I /path/to/mlpack/build/include test.cpp -l armadillo`

The above command works perfectly fine in my case (on macOS Big Sur). You might need to link mlpack also using `-l mlpack`. (I don't need to link mlpack because I have configured it in such a way that it is linked automatically.)

I hope this should work for you too. Let us know if you are still facing any issues.

**Comment by mircodemarchi** [comment]
Hi @RishabhGarg108, thank you for the suggestion. I think that the g++ that you are talking about is linked to clang++. Therefore, using g++ or using clang++ is the same. I have strict compatibility requirements for my library and I need to build it also with g++-10.

Try the command `g++ --version` and `clang++ --version`, if you didn't change macOS default settings, you will found that they are the same. 

However, both g++-10 and clang++ work in Ubuntu 20.04, so I hope there is a solution for macOS as well.

**Comment by RishabhGarg108** [comment]
> Try the command g++ --version and clang++ --version, if you didn't change macOS default settings, you will found that they are the same.

This is interesting. I never checked it before. But yes, I checked and they are indeed the same.

Well, if your requirements are strict, then I think you have to wait till this gets resolved :).

**Comment by shrit** [comment]
@mircodemarchi Would it be possible to use the command `file /path/to/libarmadillo`? and let me know the output please

**Comment by mircodemarchi** [comment]
@shrit The following is the output of `file /usr/local/lib/libarmadillo.dylib`:
```
/usr/local/lib/libarmadillo.dylib: Mach-O 64-bit dynamically linked shared library x86_64
```

**Comment by shrit** [comment]
As I thought, it seems to me that, the new macOS that is based on arm architecture is installing `x86_64` libraries.
I do have very limited knowledge of macOS, it seems that they did not recompile all of their available packages for the new architecture, but instead they allowed the installation of the libraries with the old `x86_64` architecture.
This is what is causing the issue on your machine, you are trying to link arm binaries with `x86_64` binaries which is impossible technically as they are incompatible.
To resolve the issue, I think you have two options: 1) You need to install manually `armadillo` and its dependencies (mainly openblas), and then point out to mlpack where is the location of the new library (remove the old one for safety)
Or, 2) use the mechanism of binaries translation provided by Apple.

I would vote for the first one, it will be much easier. I hope you find this helpful

Let me know what do you think, or if you need any clarification,

**Comment by shrit** [comment]
Also, I am not sure if you are using the newest mac, so it might be a different issue.  Let me know if you will face the same issue if you recompile and install everything manually.

**Comment by mircodemarchi** [comment]
I'm using a Mac mini (Late 2012) Intel core with macOS Catalina 10.15.7, therefore the reason is not the arm architecture. However, I agree with you that the problem is not directly related with mlpack but it regards armadillo library. I will try soon to build armadillo library with GPU gcc-10 and g++-10, and I let you know what happen. 

Thank you very much for the answers.

**Comment by conradsnicta** [comment]
@mircodemarchi There are pre-built versions of Armadillo on Homebrew (https://formulae.brew.sh/formula/armadillo) and MacPorts (https://ports.macports.org/port/armadillo/summary). Is using any of these giving you problems?

**Comment by mircodemarchi** [comment]
Thank you @conradsnicta for the suggestion, but unluckly pre-built versions of Armadillo are made with Clang. Instead I need the Apple GNU GCC version. 

Anyway, I resolved the Armadillo linking issue recompiling it with g++-10 (Apple GNU GCC). In particular I found 2 ways to perform this operation: 
- manually build from source: 
```
brew install openblas
wget http://sourceforge.net/projects/arma/files/armadillo-10.4.1.tar.xz
tar -xf armadillo-10.4.1.tar.xz
cd armadillo-10.4.1
cmake . -DALLOW_OPENBLAS_MACOS=ON -DCMAKE_C_COMPILER=gcc-10 -DCMAKE_CXX_COMPILER=g++-10
sudo make install
```
- homebrew build from source (not supported by homebrew, because it can provide errors but it works): 
```
brew install openblas
brew install --build-from-source --cc=gcc-10 armadillo
```

Although I solved the Armadillo linking issues, similar issues appeared for Boost library:
```
[ 19%] Linking CXX executable ../../../../bin/mlpack_bayesian_linear_regression
Undefined symbols for architecture x86_64:
  "__ZN5boost7archive17xml_iarchive_implINS0_12xml_iarchiveEEC2ERSij", referenced from:
      __ZN6mlpack4data4LoadINS_10regression24BayesianLinearRegressionEEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESB_RT_bNS0_6formatE.isra.0 in bayesian_linear_regression_main.cpp.o
  "__ZN5boost7archive17xml_oarchive_implINS0_12xml_oarchiveEEC2ERSoj", referenced from:
      __ZN6mlpack4data4SaveINS_10regression24BayesianLinearRegressionEEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESB_RT_bNS0_6formatE.isra.0 in bayesian_linear_regression_main.cpp.o
  "__ZN5boost7archive18text_iarchive_implINS0_13text_iarchiveEEC2ERSij", referenced from:
      __ZN6mlpack4data4LoadINS_10regression24BayesianLinearRegressionEEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESB_RT_bNS0_6formatE.isra.0 in bayesian_linear_regression_main.cpp.o
  "__ZN5boost7archive18text_oarchive_implINS0_13text_oarchiveEE4saveERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE", referenced from:
      __ZN5boost7archive6detail15common_oarchiveINS0_13text_oarchiveEE5vsaveERKNS0_15class_name_typeE in bayesian_linear_regression_main.cpp.o
  "__ZN5boost7archive18text_oarchive_implINS0_13text_oarchiveEEC2ERSoj", referenced from:
      __ZN6mlpack4data4SaveINS_10regression24BayesianLinearRegressionEEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESB_RT_bNS0_6formatE.isra.0 in bayesian_linear_regression_main.cpp.o
  "__ZN5boost7archive21basic_text_iprimitiveISiED2Ev", referenced from:
      __ZN5boost7archive13text_iarchiveD1Ev in bayesian_linear_regression_main.cpp.o
      __ZN5boost7archive13text_iarchiveD0Ev in bayesian_linear_regression_main.cpp.o
      __ZN6mlpack4data4LoadINS_10regression24BayesianLinearRegressionEEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESB_RT_bNS0_6formatE.isra.0 in bayesian_linear_regression_main.cpp.o
      __ZN6mlpack4data4LoadINS_10regression24BayesianLinearRegressionEEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESB_RT_bNS0_6formatE.isra.0.cold in bayesian_linear_regression_main.cpp.o
  "__ZN5boost7archive21basic_text_oprimitiveISoED2Ev", referenced from:
      __ZN5boost7archive13text_oarchiveD1Ev in bayesian_linear_regression_main.cpp.o
      __ZN5boost7archive13text_oarchiveD0Ev in bayesian_linear_regression_main.cpp.o
      __ZN6mlpack4data4SaveINS_10regression24BayesianLinearRegressionEEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESB_RT_bNS0_6formatE.isra.0 in bayesian_linear_regression_main.cpp.o
      __ZN6mlpack4data4SaveINS_10regression24BayesianLinearRegressionEEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESB_RT_bNS0_6formatE.isra.0.cold in bayesian_linear_regression_main.cpp.o
  "__ZN5boost7archive23basic_binary_iprimitiveINS0_15binary_iarchiveEcSt11char_traitsIcEE4initEv", referenced from:
      __ZN6mlpack4data4LoadINS_10regression24BayesianLinearRegressionEEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESB_RT_bNS0_6formatE.isra.0 in bayesian_linear_regression_main.cpp.o
  "__ZN5boost7archive23basic_binary_iprimitiveINS0_15binary_iarchiveEcSt11char_traitsIcEEC2ERSt15basic_streambufIcS4_Eb", referenced from:
      __ZN6mlpack4data4LoadINS_10regression24BayesianLinearRegressionEEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESB_RT_bNS0_6formatE.isra.0 in bayesian_linear_regression_main.cpp.o
  "__ZN5boost7archive23basic_binary_iprimitiveINS0_15binary_iarchiveEcSt11char_traitsIcEED2Ev", referenced from:
      __ZN5boost7archive20binary_iarchive_implINS0_15binary_iarchiveEcSt11char_traitsIcEED1Ev in bayesian_linear_regression_main.cpp.o
      __ZN5boost7archive20binary_iarchive_implINS0_15binary_iarchiveEcSt11char_traitsIcEED0Ev in bayesian_linear_regression_main.cpp.o
      __ZN5boost7archive15binary_iarchiveD1Ev in bayesian_linear_regression_main.cpp.o
      __ZN5boost7archive15binary_iarchiveD0Ev in bayesian_linear_regression_main.cpp.o
      __ZN6mlpack4data4LoadINS_10regression24BayesianLinearRegressionEEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESB_RT_bNS0_6formatE.isra.0 in bayesian_linear_regression_main.cpp.o
      __ZN6mlpack4data4LoadINS_10regression24BayesianLinearRegressionEEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESB_RT_bNS0_6formatE.isra.0.cold in bayesian_linear_regression_main.cpp.o
  "__ZN5boost7archive23basic_binary_oprimitiveINS0_15binary_oarchiveEcSt11char_traitsIcEE4initEv", referenced from:
      __ZN6mlpack4data4SaveINS_10regression24BayesianLinearRegressionEEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESB_RT_bNS0_6formatE.isra.0 in bayesian_linear_regression_main.cpp.o
  "__ZN5boost7archive23basic_binary_oprimitiveINS0_15binary_oarchiveEcSt11char_traitsIcEE4saveERKNSt7__cxx1112basic_stringIcS4_SaIcEEE", referenced from:
      __ZN5boost7archive6detail15common_oarchiveINS0_15binary_oarchiveEE5vsaveERKNS0_15class_name_typeE in bayesian_linear_regression_main.cpp.o
  "__ZN5boost7archive23basic_binary_oprimitiveINS0_15binary_oarchiveEcSt11char_traitsIcEEC2ERSt15basic_streambufIcS4_Eb", referenced from:
      __ZN6mlpack4data4SaveINS_10regression24BayesianLinearRegressionEEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESB_RT_bNS0_6formatE.isra.0 in bayesian_linear_regression_main.cpp.o
  "__ZN5boost7archive23basic_binary_oprimitiveINS0_15binary_oarchiveEcSt11char_traitsIcEED2Ev", referenced from:
      __ZN5boost7archive20binary_oarchive_implINS0_15binary_oarchiveEcSt11char_traitsIcEED1Ev in bayesian_linear_regression_main.cpp.o
      __ZN5boost7archive20binary_oarchive_implINS0_15binary_oarchiveEcSt11char_traitsIcEED0Ev in bayesian_linear_regression_main.cpp.o
      __ZN5boost7archive15binary_oarchiveD1Ev in bayesian_linear_regression_main.cpp.o
      __ZN5boost7archive15binary_oarchiveD0Ev in bayesian_linear_regression_main.cpp.o
      __ZN6mlpack4data4SaveINS_10regression24BayesianLinearRegressionEEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESB_RT_bNS0_6formatE.isra.0 in bayesian_linear_regression_main.cpp.o
      __ZN6mlpack4data4SaveINS_10regression24BayesianLinearRegressionEEEbRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESB_RT_bNS0_6formatE.isra.0.cold in bayesian_linear_regression_main.cpp.o
ld: symbol(s) not found for architecture x86_64
collect2: error: ld returned 1 exit status
make[2]: *** [bin/mlpack_bayesian_linear_regression] Error 1
make[1]: *** [src/mlpack/methods/bayesian_linear_regression/CMakeFiles/mlpack_bayesian_linear_regression.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
...
```

Therefore I tried to build boost library from source with Apple GNU GCC following the same two alternatives:
- manually build from source: 
```
brew install icu4c
wget https://boostorg.jfrog.io/artifactory/main/release/1.75.0/source/boost_1_75_0.tar.gz
tar -xf boost_1_75_0.tar.gz
cd boost_1_75_0
CC=gcc-10 CXX=g++-10 ./bootstrap.sh --with-icu=/usr/local/opt/icu4c --without-libraries=python,mpi
./b2 headers
echo "using darwin : : g++-10 ;\n" > user-config.jam
./b2 -d2 -j8 --layout=tagged-1.66 --user-config=user-config.jam -sNO_LZMA=1 -sNO_ZSTD=1 install threading=multi,single
```
- homebrew build from source: not supported and it doesn't work :/

After building manually from sources the Boost library, I was able to compile also mlpack with Apple GNU GCC following the steps described [here](https://github.com/mlpack/mlpack/issues/2934#issuecomment-830916125). However, I would prefer to install Boost and mlpack through Homebrew. For this reason, I opened an [issue](https://github.com/Homebrew/homebrew-core/issues/76645) in the official [Homebrew/homebrew-core](https://github.com/Homebrew/homebrew-core) repository and I asked if there is a way to use Homebrew to build Boost library with Apple GNU GCC. However, as I mentioned before, Homebrew doesn't support building from source, and especially not when another compiler is specified. 

In conclusion, there is no way to install mlpack for Apple GNU GCC through Homebrew, because Homebrew doesn't support custom build. The only way to install mlpack for Apple GNU GCC is manually build Armadillo, Boost and then mlpack from source.
